### PR TITLE
[CDAP-14284] Adds base implementation to handle different types of nodes in React

### PR DESCRIPTION
**Pull request by ajainarayanan**
**Things done**
- Adds an `AbstractNode` to handle all the common actions required for rendering a node
- Adds `Source, Transform, Sink, Alert, Error` node types.
- Styling changes to resemble endpoints in existing implenentation

**Things to do**
- Add splitter, condition and action node types
- Streamline styling for nodes.
- Seperate out jsplumb settings and styles required for nodes. Right now it is not ideal 
Build: 
https://builds.cask.co/browse/CDAP-UDUT477

### CODE REVIEW

File: cdap-ui/app/cdap/components/DAG/Nodes/AlertPublisherNode/index.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  '&.drag-hover': {
+    backgroundColor: 'rgba(255, 186, 1, 0.1)',
+  },
+});
+interface IAlertPublisherProps extends IAbstractNodeProps<typeof styles> {}
+class AlertPublisherComponent extends AbstractNode<IAlertPublisherProps> {
+  public type = 'alertpublisher';
+
+  public checkForValidIncomingConnection = (connObj) => {
+    if (connObj.connection.target.getAttribute('data-node-type') !== 'alertpublisher') {
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/DAG/Nodes/AlertPublisherNode/index.tsx:32]: i'm not sure i understand this condition, why are we checking the target?

File: cdap-ui/app/cdap/components/DAG/Nodes/TransformNode/index.tsx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+      ) === -1
+    );
+  };
+
+  public render() {
+    const {
+      classes,
+      config: { showAlert = false, showError = false },
+    } = this.props;
+    const RegularEndpoint = () => (
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/DAG/Nodes/TransformNode/index.tsx:72]: should there be an abstract component for these endpoints as well?

File: cdap-ui/app/cdap/components/DAG/Nodes/AlertPublisherNode/index.tsx (hunk 1 above)
Comment (ajainarayanan) [reply cdap-ui/app/cdap/components/DAG/Nodes/AlertPublisherNode/index.tsx:32]: - Every node will have `checkForValidIncomingConnection` condition to determine if the current connection is valid or not.
- We register this callback from each node in the `DAGRenderer` [init node function ](https://github.com/cdapio/cdap/blob/feature/ui-CDAP-14284-2/cdap-ui/app/cdap/components/DAG/DAGRenderer.tsx#L248-L250).
- When a `connection` event is fired when user connects two nodes we call all the callbacks from the nodes. 
- It is basically a [reduce on all the callbacks](https://github.com/cdapio/cdap/blob/feature/ui-CDAP-14284-2/cdap-ui/app/cdap/components/DAG/DAGRenderer.tsx#L253-L255) and the eventual reduce value should be `true` for the connection to be drawn. It it returns false then one of the nodes involved in the connection rejects this connection so we don't render the connection.

The advantage of this is having such custom logic specific to node within the node.

Hope I made sense.

~Now that I think about it I think I can add this function as a static function and have it registered as not per instance of the node. Will add that change.~

EDIT:
The connection validation event handlers are still part of the instance but we register them based on node types and not node instances. This prevents adding a handler for each instance of the node and rather adds one handler per node type.

File: cdap-ui/app/cdap/components/DAG/Nodes/TransformNode/index.tsx (hunk 2 above)
Comment (ajainarayanan) [reply cdap-ui/app/cdap/components/DAG/Nodes/TransformNode/index.tsx:72]: I didn't perceive it that way. Right now all nodes have a generic endpoint and if the node needs to change in their specific implementation then they can override this function (This is becoming a class function in AbstractNode in my next PR which will be overridden by child nodes if need be).

**Review by elfenheart — APPROVED** [review]
👍 LGTM

**Pull request merged**